test/async-action.ts

```diff
@@ -214,6 +214,192 @@ test("it should support logging", t => {
     }, 10)
 })
 
+test("it should support asyncActionWithCancel(cb, type) as decorator for have-value variable", t => {
+    const values: any[] = []
+
+    mobx.useStrict(true)
+
+    class X {
+        @mobx.observable a = 1;
+
+        @utils.asyncActionWithCancel(() => {}, 'takeLatest')
+        *f(initial: number) {
+            this.a = initial // this runs in action
+            try {
+                this.a = yield delay(100, 5, true) // and this as well!
+                yield delay(100, 0)
+                this.a = 4
+            } catch (e) {
+                this.a = e
+            }
+            return this.a
+        }
+    }
+
+    const x = new X()
+    mobx.reaction(() => x.a, v => values.push(v), { fireImmediately: true })
+
+    setTimeout(() => {
+        // TODO: mweh on any cast...
+        ;(x.f(/*test binding*/ 2) as any).then((v: number) => {
+            // note: ideally, type of v should be inferred..
+            t.is(v, 5)
+            t.deepEqual(values, [1, 2, 5])
+            t.is(x.a, 5) // correct instance modified?
+            t.end()
+        })
+    }, 10)
+})
+
+test("it should support asyncActionWithCancel(cb, type) as decorator for no-value variable", t => {
+    const values: any[] = []
+
+    mobx.useStrict(true)
+
+    class X {
+        @mobx.observable a = 1;
+
+        @utils.asyncActionWithCancel(() => {}, 'takeLatest')
+        noValueParam
+    }
+
+    const x = new X()
+    x.noValueParam = function *(initial: number) {
+        this.a = initial // this runs in action
+        try {
+            this.a = yield delay(100, 5, true) // and this as well!
+            yield delay(100, 0)
+            this.a = 4
+        } catch (e) {
+            this.a = e
+        }
+        return this.a
+    }
+
+    mobx.reaction(() => x.a, v => values.push(v), { fireImmediately: true })
+
+    setTimeout(() => {
+        // TODO: mweh on any cast...
+        ;(x.noValueParam(/*test binding*/ 2) as any).then((v: number) => {
+            // note: ideally, type of v should be inferred..
+            t.is(v, 5)
+            t.deepEqual(values, [1, 2, 5])
+            t.is(x.a, 5) // correct instance modified?
+            t.end()
+        })
+    }, 10)
+})
+
+test("it should support canceling the effects when call canceler", t => {
+    mobx.useStrict(true)
+    const values: any[] = []
+    const x = mobx.observable({ a: 1 })
+    mobx.reaction(() => x.a, v => values.push(v), { fireImmediately: true })
+
+    let canceler
+    const f = utils.asyncActionWithCancel(cancel => canceler = cancel)(function*(initial: number) {
+        x.a = initial // this runs in action
+        try {
+            x.a = yield delay(1000, 5, true) // and this as well!
+        } catch (e) {
+            x.a = e
+        }
+        return x.a
+    })
+
+    setTimeout(() => {
+        f(2)
```

Comment:
When cancelling, what happens to the `Promise` return from `f(2)`? Does it reject with some special cancellation token? (That is what I would expect)

Reply:
Yes.
Since the canceler is base on the **[generator.prototype.return](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Generator/return)**, it is possible to return some special token for reject which is the string 'cancel' in my pr.

something like this
```
generator.return({ error: 'cancel' })
const ret = generator.next() // { done: true, value: { error: 'cancel'} } 

if (ret.done) {
   reject(ret.value)
}
```